File: problems_001to025/problem_004/solution.py
import os
import sys

PROJECT_PATH = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.append(PROJECT_PATH)

from measurement import Stopwatch
from toolbox import positive_divisors


class Solution(Stopwatch):
    def __init__(self):
        super().__init__()

    def execute(self):
        answer = None
        for num in range(997, 99, -1):
            digits = [digit for digit in str(num)]
            digits.reverse()
            reversed_num = int(''.join(digits))
            palindrome = num*1000 + reversed_num
            divisors = positive_divisors(palindrome)
            valid_divisors = [divisor for divisor in divisors if divisor < 1000 and divisor >= 100]
            quotients = [palindrome // divisor for divisor in valid_divisors]
            valid_quotients = [quotient for quotient in quotients if quotient < 1000 and quotient >= 100]
            if len(valid_quotients) > 0:
                answer = palindrome
                break
        return answer

if __name__ == '__main__':
    soln = Solution()
    result = soln.execute()
    print(result)

File: problems_001to025/problem_004/solution.js
const { positiveDivisors } = require('../../lib/toolbox');

const solution = () => {
  let answer = null;

  for (let num = 997; num >= 100; num -= 1) {
    const reversedNum = Number(String(num).split('').reverse().join(''));
    const palindrome = num * 1000 + reversedNum;

    const divisors = positiveDivisors(palindrome);
    const validDivisors = divisors.filter((divisor) => (divisor >= 100 && divisor <= 999));

    const quotients = validDivisors.map((divisor) => (palindrome / divisor));
    const validQuotients = quotients.filter((quotient) => (quotient >= 100 && quotient <= 999));

    if (validQuotients.length > 0) {
      answer = palindrome;
      break;
    }
  }

  return answer;
};

if (process.env.NODE_ENV !== 'test') {
  // eslint-disable-next-line no-console
  console.log(solution());
}

module.exports = solution;
